Arnej/fix logctl coredump

arnej27959 opened the pull request:
@havardpe please review and merge

arnej27959 committed:
avoid parallel-make problem

arnej27959 committed:
ensure returned Component refers to valid memory

* we have seen some core dumps in logctl when using
  the mmap'ed levels inside a Component.  Add extra
  checks to ensure we stay inside the actual file.

havardpe reviewed (commented):
conclusion: the undefined behavior is probably in our favor. the check whether the start of the next line is inside the file will probably save us. the crash was probably caused by a new line written by another program that crosses the last mapped page boundary.

arnej27959 committed:
avoid possible memory overrun in strchr()

* add a loop to look for newlines that takes the known
  file size into account so we don't risk accessing
  outside the valid mapped area.
* also, add a log warning (acting as explanation for
  this-should-not-happen handling code).

havardpe reviewed (commented):
I agree that the undefined mmap behavior seems to be ok for now. I did however re-check the nn vs end pointer logic and need additional reassurance why this is correct.

arnej27959 committed:
fix bug found in review
